### PR TITLE
I/O Performance Measurement: Add I/O Performance Monitor

### PR DESCRIPTION
**Pull request by tobonex**
Implements I/O performance measurement feature. It counts number of transmitted bytes or other events that happen in various interfaces. Currently measured interfaces: IPC, IDC, SNDW, HDA, SSP, DMIC. 

There are some interfaces that are defined in Reference Firmware that apparently are not used directly here: USB, I2C, I3C, UART, CSI_2 , DTF. But I could not confirm this as of yet.

There are some other minor issues I will mention in comments.

### CODE REVIEW

File: src/ipc/ipc-common.c [hunk 1]
```diff
@@ -282,6 +282,7 @@ int ipc_init(struct sof *sof)
 	list_init(&sof->ipc->comp_list);
 
 #ifdef CONFIG_SOF_TELEMETRY_IO_PERFORMANCE_MEASUREMENTS
+	io_perf_monitor_init();
```
Comment (tobonex) [inline src/ipc/ipc-common.c:285]: I need to find a nice place for this init. This has to go before ipc_init. Not sure if we should consider this platform specific (it does not use memory window like previous performance meas. features).

File: src/debug/telemetry/performance_monitor.c [hunk 2]
```diff
@@ -397,3 +404,236 @@ int performance_monitor_init(void)
...
+	return 0;
+}
+
+void io_perf_monitor_update_data(struct io_perf_data_item *slot_id, uint32_t increment)
+{
+	if (!slot_id)
+		return;
+
+	struct io_perf_monitor_ctx *self = &perf_monitor_ctx;
+	k_spinlock_key_t key = k_spin_lock(&self->lock);
```
Comment (tobonex) [inline src/debug/telemetry/performance_monitor.c:619]: I'm not sure about this, this lock was there in reference (not sure what kind of a lock it was), but I think this has potential to disrupt the measured interfaces with a spinlock here. The entries to which we write are just statically defined memory, assigned to a module from performance monitor, and released by the module itself, so I don't think the global lock is needed.

**Review by tobonex — COMMENTED** [review]
Some comments from me.

File: src/debug/telemetry/performance_monitor.c (hunk 2 above)
Comment (tobonex) [reply src/debug/telemetry/performance_monitor.c]: CI answered my own question, smokes did not pass. Both this and release slot function cannot be locked. io_perf_monitor_release_slot has locks in the bitmap functions, so it should be safe.

File: src/ipc/ipc-common.c (hunk 1 above)
Comment (lgirdwood) [reply src/ipc/ipc-common.c:285]: Yeah, it almost like we need an entry point for certain system feature/services and we put them in order withing a global services_init function.

File: src/debug/telemetry/Kconfig [hunk 3]
```diff
@@ -18,3 +18,12 @@ config SOF_TELEMETRY_PERFORMANCE_MEASUREMENTS
 	  Performance records are stored in the limited number of slots in Memory Window 3,
 	  so only a certain number (PERFORMANCE_DATA_ENTRIES_COUNT) of components can be measured.
 
+config SOF_TELEMETRY_IO_PERFORMANCE_MEASUREMENTS
+	bool "enable performance measurements"
```
Comment (kv2019i) [inline src/debug/telemetry/Kconfig:22]: Maybe add the "enable I/O performance measurements" to the short string as well as this is specifically I/O perf

File: src/debug/telemetry/performance_monitor.c [hunk 4]
```diff
@@ -397,3 +404,230 @@ int performance_monitor_init(void)
...
+	return 0;
+}
+
+void io_perf_monitor_update_data(struct io_perf_data_item *slot_id, uint32_t increment)
+{
+	if (!slot_id)
+		return;
+
+	struct io_perf_monitor_ctx *self = &perf_monitor_ctx;
+	//k_spinlock_key_t key = k_spin_lock(&self->lock);
```
[the remark this answers is not in the document]
Comment (kv2019i) [reply src/debug/telemetry/performance_monitor.c:616]: Given "data" is 64bit, atomics won't work, but it would seem locking is probably not needed given a single slot is only used for one interface.

File: src/debug/telemetry/performance_monitor.c [hunk 5]
```diff
@@ -397,3 +404,230 @@ int performance_monitor_init(void)
...
+	return &self->io_perf_data[idx];
+}
+
+int io_perf_monitor_release_slot(struct io_perf_data_item *item)
+{
+	struct io_perf_monitor_ctx *self = &perf_monitor_ctx;
+
+	item->is_removed = true;
+
+	int idx = (item - self->io_perf_data) / sizeof(*item);
```
Comment (kv2019i) [inline src/debug/telemetry/performance_monitor.c:467]: Wouldn't this be one place where locking is needed?

**Review by kv2019i — COMMENTED** [review]
Multiple minor comments. The locking is a bit hard to follow, but otherwise looks ok.

File: src/debug/telemetry/performance_monitor.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (tobonex) [reply src/debug/telemetry/performance_monitor.c]: Yes, if only one instance uses a specific slot this should be ok. I'm not 100% sure if this was the original design but it is a thing now. I did add a comment in .h now to make this apaprent.

File: src/debug/telemetry/performance_monitor.c (hunk 5 above)
Comment (tobonex) [reply src/debug/telemetry/performance_monitor.c]: I added lock for this line alone, we'll see if we get any problems in smoke tests. This function did deadlock when i locked all the code here, but it's probably because bitmap has it's own locks too.

File: src/debug/telemetry/performance_monitor.c [hunk 6]
```diff
@@ -38,6 +38,13 @@ struct perf_bitmap {
 	size_t size;
 };
 
+struct io_perf_monitor_ctx {
+	struct k_spinlock lock;
+	enum ipc4_perf_measurements_state_set state;
+	struct perf_bitmap io_performance_data_bitmap;//??????
```
[the remark this answers is not in the document]
Comment (tobonex) [reply src/debug/telemetry/performance_monitor.c]: Ah, I do use it. It may be unnecessary though, as this only points to static memory.

File: src/include/sof/debug/telemetry/performance_monitor.h [hunk 7]
```diff
@@ -87,4 +89,154 @@ int enable_performance_counters(void);
...
+	USB_ID = 6,
+	GPIO_ID = 7,
+	I2C_ID = 8,
+	I3C_ID = 9,
+	I3C_INTERRUPT_ID = 10,
+	UART_ID = 11,
+	SPI_ID = 12,
+	CSI_2_ID = 13,
+	DTF_ID = 14,
+	INVALID_ID = 0xFF
```
[the remark this answers is not in the document]
Comment (tobonex) [reply src/include/sof/debug/telemetry/performance_monitor.h]: Added IO_PERF_

File: src/debug/telemetry/performance_monitor.c [hunk 8]
```diff
@@ -397,3 +404,225 @@ int performance_monitor_init(void)
...
+}
+
+static int
+io_perf_monitor_disabled(struct io_perf_monitor_ctx *self)
+{
+	return 0;
+}
+
+static int
+io_perf_monitor_stopped(struct io_perf_monitor_ctx *self)
```
Comment (lyakh) [inline src/debug/telemetry/performance_monitor.c:516]: this never returns errors, so can be `void`. Besides the name sounds like it's checking whether the monitor has stopped, while this isn't what this function is doing

File: src/debug/telemetry/performance_monitor.c [hunk 9]
```diff
@@ -397,3 +404,225 @@ int performance_monitor_init(void)
...
+		ret = io_perf_monitor_disabled(self);
+		break;
+	case IPC4_PERF_MEASUREMENTS_STOPPED:
+		ret = io_perf_monitor_stopped(self);
+		break;
+	case IPC4_PERF_MEASUREMENTS_STARTED:
+		ret = io_perf_monitor_started(self);
+		break;
+	case IPC4_PERF_MEASUREMENTS_PAUSED:
+		ret = io_perf_monitor_paused(self);
```
Comment (lyakh) [inline src/debug/telemetry/performance_monitor.c:562]: I'd call all these functions with infinitive verb forms as `disable()`, `stop()`, etc.

File: src/debug/telemetry/performance_monitor.c [hunk 10]
```diff
@@ -397,3 +404,225 @@ int performance_monitor_init(void)
...
+	return self->state;
+}
+
+int io_perf_monitor_init_data(struct io_perf_data_item **slot_id,
+			      struct io_perf_data_item *init_data)
+{
+	if (!slot_id)
+		return IPC4_ERROR_INVALID_PARAM;
+
+	struct io_perf_monitor_ctx *io_perf_monitor = &perf_monitor_ctx;
```
Comment (lyakh) [inline src/debug/telemetry/performance_monitor.c:588]: you usually call this `self`

File: src/debug/telemetry/performance_monitor.c [hunk 11]
```diff
@@ -397,3 +404,225 @@ int performance_monitor_init(void)
...
+
+void io_perf_monitor_update_data(struct io_perf_data_item *slot_id, uint32_t increment)
+{
+	if (!slot_id)
+		return;
+
+	struct io_perf_monitor_ctx *self = &perf_monitor_ctx;
+
+	/* this does not need a lock if each perf slot has only one user */
+	if (IPC4_PERF_MEASUREMENTS_STARTED == self->state && NULL != slot_id)
```
Comment (lyakh) [inline src/debug/telemetry/performance_monitor.c:615]: you already checked `slot_id`

File: src/audio/base_fw.c [hunk 12]
```diff
@@ -467,6 +467,45 @@ static int global_perf_data_get(uint32_t *data_off_size, char *data)
 #endif
 }
 
+static int io_global_perf_state_get(uint32_t *data_off_size, char *data)
```
Comment (lyakh) [inline src/audio/base_fw.c:470]: `size_t`? `char` is also a bit confusing for `data`, why not use the appropriate `enum`

File: src/audio/base_fw.c [hunk 13]
```diff
@@ -467,6 +467,45 @@ static int global_perf_data_get(uint32_t *data_off_size, char *data)
...
+	*data = io_perf_monitor_get_state();
+	*data_off_size = sizeof(enum ipc4_perf_measurements_state_set);
+
+	return IPC4_SUCCESS;
+#else
+	return IPC4_UNAVAILABLE;
+#endif
+}
+
+static int io_global_perf_data_get(uint32_t *data_off_size, char *data)
```
Comment (lyakh) [inline src/audio/base_fw.c:482]: why `char`? what does it have to do with characters / strings?

File: src/ipc/ipc-common.c [hunk 14]
```diff
@@ -317,6 +336,11 @@ static enum task_state ipc_do_cmd(void *data)
 {
 	struct ipc *ipc = data;
 
+#ifdef CONFIG_SOF_TELEMETRY_IO_PERFORMANCE_MEASUREMENTS
+	/* Increment performance counters */
+	io_perf_monitor_update_data(ipc->io_perf_in_msg_count, 1);
```
Comment (lyakh) [inline src/ipc/ipc-common.c:342]: why IPC commands are counted for performance? In some use-cases there are many IPCs, in others none, like plain playback or capture?

File: src/idc/idc.c [hunk 15]
```diff
@@ -481,6 +483,11 @@ static void idc_complete(void *data)
 	uint32_t type = iTS(idc->received_msg.header);
 	k_spinlock_key_t key;
 
+#ifdef CONFIG_SOF_TELEMETRY_IO_PERFORMANCE_MEASUREMENTS
+	/* Increment performance counters */
+	io_perf_monitor_update_data(idc->io_perf_out_msg_count, 1);
```
Comment (lyakh) [inline src/idc/idc.c:488]: same here, how does the number of IDCs describe IO performance?

File: src/audio/dai-zephyr.c [hunk 16]
```diff
@@ -438,6 +444,47 @@ int dai_common_new(struct dai_data *dd, struct comp_dev *dev,
...
+		perf_type = IO_PERF_HDA_ID;
+		break;
+
+	default:
+		perf_type = IO_PERF_INVALID_ID;
+	}
+	if (dai_cfg->direction == SOF_IPC_STREAM_PLAYBACK)
+		perf_dir = IO_PERF_INPUT_DIRECTION;
+	else
+		perf_dir = IO_PERF_OUTPUT_DIRECTION;
```
Comment (lyakh) [inline src/audio/dai-zephyr.c:474]: is this correct? intuitively I'd associate playback with output and capture with input? Particularly since above you set the type according to the DAI, so it doesn't seem to be related to the host interface

File: src/audio/dai-zephyr.c [hunk 17]
```diff
@@ -438,6 +444,47 @@ int dai_common_new(struct dai_data *dd, struct comp_dev *dev,
...
+						      cpu_get_id(),
+						      perf_dir,
+						      IO_PERF_POWERED_UP_ENABLED,
+						      IO_PERF_D0IX_POWER_MODE,
+						      0, 0, 0 };
+		io_perf_monitor_init_data(&dd->io_perf_bytes_count, &init_data);
+	}
+#endif
+
 	return 0;
```
Comment (lyakh) [inline src/audio/dai-zephyr.c:488]: also in the `INVALID` case?

Comment (tobonex) [reply src/audio/dai-zephyr.c:528]: Yes, we should treat performance as something optional. Perf setup failing should not make everything else fail

File: src/audio/dai-zephyr.c (hunk 16 above)
Comment (tobonex) [reply src/audio/dai-zephyr.c:512]: Yes, you are right. Didn't know how all of this worked back then. Changed it.

File: src/ipc/ipc-common.c (hunk 14 above)
Comment (tobonex) [reply src/ipc/ipc-common.c:342]: In ref. FW this feature did also count some events like IPC IDC and some interrupts.

File: src/audio/base_fw.c (hunk 13 above)
Comment (tobonex) [reply src/audio/base_fw.c:482]: A better question would be: why all the other handlers have this kind of interface

File: src/debug/telemetry/performance_monitor.c [hunk 18]
```diff
@@ -397,3 +404,225 @@ int performance_monitor_init(void)
...
+
+	item->is_removed = true;
+
+	key = k_spin_lock(&self->lock);
+	idx = (item - self->io_perf_data) / sizeof(*item);
+	k_spin_unlock(&self->lock, key);
+
+	ret = perf_bitmap_clearbit(&self->io_performance_data_bitmap, idx);
+	if (ret < 0)
+		return ret;
```
[the remark this answers is not in the document]
Comment (tobonex) [reply src/debug/telemetry/performance_monitor.c:476]: This only fails if idx is wrong - "free" will also fail in this case.

[the remark this answers is not in the document]
Comment (lyakh) [reply src/debug/telemetry/performance_monitor.c:476]: and `idx` can only be wrong if `item` is wrong, and you've already used it above to set `item->is_removed = true`. So, if `item` can really be wrong, can we check it first before modifying anything? And if it cannot be wrong, then we should probably handle "impossible" cases differently here - either just print a big scary error message or even use an `assert()`

File: src/debug/telemetry/performance_monitor.c [hunk 19]
```diff
@@ -397,3 +404,225 @@ int performance_monitor_init(void)
...
+	ret = perf_bitmap_free(&performance_data_bitmap, idx);
+	if (ret < 0)
+		return ret;
+
+	return 0;
```
Comment (lyakh) [inline src/debug/telemetry/performance_monitor.c:481]: btw, this should've been
```
return perf_bitmap_free(&performance_data_bitmap, idx);
```
but if you agree to use asserts, then something like
```
ret = perf_bitmap_free(&performance_data_bitmap, idx);
assert(!ret);
return ret;
```
You could also consider `return 0` but then if built with asserts disabled, the compiler might complain about assigned but not used `ret`

File: src/debug/telemetry/performance_monitor.c (hunk 18 above)
[the remark this answers is not in the document]
Comment (tobonex) [reply src/debug/telemetry/performance_monitor.c:483]: Yeah, there's no check for null., and there should be one, added. As `item` is within a certain memory range we could check if it is within it also.

**Review by lgirdwood — COMMENTED** [review]
@tobonex I assume when IO is enabled, the the data will be gathered on every IO action and will be reset when D3. Btw, it occurs to me now we are gathering a lot of good data than we can also additionally now dump via the Zephyr shell. @kv2019i

File: src/debug/telemetry/performance_monitor.c [hunk 20]
```diff
@@ -433,7 +433,8 @@ int io_perf_monitor_init(void)
 }
 
 /* init performance monitor using Zephyr */
-SYS_INIT(io_perf_monitor_init, APPLICATION, CONFIG_APPLICATION_INIT_PRIORITY);
+/* IMPORTANT , this init is now in IPC init, temporarily */
```
[the remark this answers is not in the document]
Comment (tobonex) [reply src/debug/telemetry/performance_monitor.c:436]: Well, this is connected to the io_perf_monitor_init comment at the top, not sure if we have a nice place for the init. Leaving it near ipc_init works, but it's a bit ugly. 

If we don't have a better place for this init, I will just leave it near ipc_init for now, and merge this temporary commit to another, or rename it.

File: src/include/sof/debug/telemetry/performance_monitor.h [hunk 21]
```diff
@@ -138,4 +140,156 @@ static inline void disable_performance_counters(void) {}
...
+	 * HD/A      - 5 - Bytes - Counter of bytes transferred over HD/A interface
+	 * USB       - 6 - Bytes - Counter of bytes transferred over USB interface
+	 * GPIO      - 7 - Count - Counter of GPIO interrupts or triggers
+	 * I2C       - 8 - Bytes - Counter of bytes transferred over I2C interface
+	 * I3C       - 9 - Bytes - Counter of bytes transferred over I3C interface
+	 * I3C interrupt - 10 - Bytes - Counter of I3C interrupts
+	 * UART      - 11 - Bytes - Counter of bytes transferred over UART interface
+	 * SPI       - 12 - Bytes - Counter of bytes transferred over SPI interface
+	 * CSI-2     - 13 - Bytes - Counter of bytes transferred over CSI-2 interface
+	 * DTF       - 14 - Bytes - Counter of bytes transferred over DTF interface
```
[the remark this answers is not in the document]
Comment (tobonex) [reply src/include/sof/debug/telemetry/performance_monitor.h:177]: The list itself is taken from ref. FW. I couldn't get info on which interfaces are actually used, so we have counters for those that are certain and on SOF side: IPC, IDC, DAI (dmic, i2s, sndw, hda),

File: src/debug/telemetry/Kconfig [hunk 22]
```diff
@@ -18,3 +18,12 @@ config SOF_TELEMETRY_PERFORMANCE_MEASUREMENTS
 	  Performance records are stored in the limited number of slots in Memory Window 3,
 	  so only a certain number (PERFORMANCE_DATA_ENTRIES_COUNT) of components can be measured.
 
+config SOF_TELEMETRY_IO_PERFORMANCE_MEASUREMENTS
+	bool "enable I/O performance measurements"
+	default n
```
Comment (kv2019i) [inline src/debug/telemetry/Kconfig:23]: "default n" is not needed for booleans -> https://docs.zephyrproject.org/latest/build/kconfig/tips.html#style-recommendations-and-shorthands

File: src/debug/telemetry/performance_monitor.c [hunk 23]
```diff
@@ -38,6 +38,13 @@ struct perf_bitmap {
 	size_t size;
```
Comment (kv2019i) [inline src/debug/telemetry/performance_monitor.c:38]: typo in commit: Implemenbts->Implements

File: src/ipc/ipc-common.c (hunk 1 above)
Comment (kv2019i) [reply src/ipc/ipc-common.c:285]: @tobonex sof/src/init/init.c ?

**Review by kv2019i — APPROVED** [review]
A few minor comments, but nothing blocking merge.

File: src/ipc/ipc-common.c (hunk 1 above)
Comment (tobonex) [reply src/ipc/ipc-common.c]: @kv2019i Moved it to primary_core_init in init.c, seems to work. Thanks!

File: src/debug/telemetry/performance_monitor.c [hunk 24]
```diff
@@ -38,6 +38,13 @@ struct perf_bitmap {
...
 };
 
+struct io_perf_monitor_ctx {
+	struct k_spinlock lock;
```
[the remark this answers is not in the document]
Comment (tobonex) [reply src/debug/telemetry/performance_monitor.c:42]: Well, this feature will probably be used in debug and test scenarios only. Also only setup and release are locked (used on module init and deletion), so the overhead is not that big i think.

File: src/debug/telemetry/performance_monitor.c [hunk 25]
```diff
@@ -397,3 +404,225 @@ int performance_monitor_init(void)
...
+	if (ret < 0)
+		return NULL;
+	/* ref. FW did not set the bits, but here we do it to not have to use
+	 * isFree() check that the bitarray does not provide yet. Instead we will use isClear
+	 * ,and always set bit on bitmap alloc.
+	 */
+
+	ret = perf_bitmap_setbit(&self->io_performance_data_bitmap, idx);
+	if (ret < 0)
+		return NULL;
```
[the remark this answers is not in the document]
Comment (tobonex) [reply src/debug/telemetry/performance_monitor.c:451]: If setbit fails, free also fails (wrong idx)

**Review by gbernatxintel — APPROVED** [review]

**Review by softwarecki — APPROVED** [review]

**Comment by lgirdwood** [comment]
Lot of failures, could have been lab power off event. rerun CI.

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by tobonex** [comment]
Some CI fails are due to newly added asserts failing. I'll take a look.

File: src/audio/dai-zephyr.c (hunk 17 above)
Comment (lyakh) [reply src/audio/dai-zephyr.c:528]: ok, then maybe at least print a warning in the `default` case at line 509 above - not a blocker, can be a follow up patch

**Review by lgirdwood — APPROVED** [review]
